Use correct accuracy for the RR domain

certik opened the pull request:
We use two extra guard digits from the maximum precision of all coefficients.

Previously:

```
In [1]: simplify(1e-14*I)
Out[1]: 0
```

Now:

```
In [1]: simplify(1e-14*I)
Out[1]: 1.0e-14⋅ⅈ
```

Fixes #9398

certik committed:
Use correct accuracy for the RR domain

We use two extra guard digits from the maximum precision of all coefficients.

Previously:

In [1]: simplify(1e-14*I)
Out[1]: 0

Now:

In [1]: simplify(1e-14*I)
Out[1]: 1.0e-14⋅ⅈ

Fixes #9398

certik commented:
A test still needs to be written. I am posting it here without tests to see if this is the right fix.

certik commented:
Only 3 tests in the polynomial module fail:
```
_________ sympy/polys/tests/test_constructor.py:test_construct_domain __________
  File "/home/ondrej/repos/sympy/sympy/polys/tests/test_constructor.py", line 17, in test_construct_domain
    assert construct_domain([3.14, 1, S(1)/2]) == (RR, [RR(3.14), RR(1.0), RR(0.5)])
AssertionError
________________________________________________________________________________
_____________ sympy/polys/tests/test_polytools.py:test_Poly__new__ _____________
  File "/home/ondrej/repos/sympy/sympy/polys/tests/test_polytools.py", line 349, in test_Poly__new__
    assert Poly(x**2 + x + 1.0).get_domain() == RR
AssertionError
________________________________________________________________________________
___________ sympy/polys/tests/test_polytools.py:test_Poly_get_domain ___________
  File "/home/ondrej/repos/sympy/sympy/polys/tests/test_polytools.py", line 563, in test_Poly_get_domain
    assert Poly(0.2*x).get_domain() == RR
AssertionError
```
That's caused by the fact that now the generated `RR` has a different precision than the default `RR`, e.g.:
```
In [1]: Poly(0.2*x)
Out[1]: Poly(0.2*x, x, domain='RR')

In [2]: Poly(0.2*x).get_domain()
Out[2]: RR

In [3]: Poly(0.2*x).get_domain() == RR
Out[3]: False
```
It seems to me that the tests are actually wrong, and you have to test for `isinstance(RieldField)` rather than the default `RR`.

certik committed:
Fix tests to pass

The failures were caused by the fact that now the generated RR has a different
precision than the default RR, e.g.:

In [1]: Poly(0.2*x)
Out[1]: Poly(0.2*x, x, domain='RR')

In [2]: Poly(0.2*x).get_domain()
Out[2]: RR

In [3]: Poly(0.2*x).get_domain() == RR
Out[3]: False

So we test for isinstance of RealField instead.

certik commented:
I fixed the tests, now they all pass, at least in the polynomials module.

certik commented:
Few more tests that fail elsewhere:
```
____________________ sympy/core/tests/test_arit.py:test_Mod ____________________
  File "/home/ondrej/repos/sympy/sympy/core/tests/test_arit.py", line 1542, in test_Mod
    assert Mod(5.0*x, 0.1*y) == 0.1*Mod(50*x, y)
AssertionError

____________ sympy/simplify/tests/test_simplify.py:test_issue_7263 _____________
  File "/home/ondrej/repos/sympy/sympy/simplify/tests/test_simplify.py", line 232, in test_issue_7263
    673.447451402970) < 1e-15
AssertionError
```

certik commented:
The second failure is almost definitely a bug in the test --- if the correct answer 673.447451402970 is only given with 12 digits after the decimal point, you can't test for accuracy of 15 digits...

certik committed:
Lower the accuracy for test_issue_7263()

The correct answer 673.447451402970 only has 12 digits after the decimal point,
so the maximum accuracy that we can test is 1e-12, not 1e-15.

certik commented:
As far as the other failure, `Mod(5.0*x, 0.1*y) == 0.1*Mod(50*x, y)`, I am not sure. It simply stopped simplifying, but I don't know if we want it to do it or not.

smichr commented:
I think [4] makes most sense:

```python
>>> Mod(5.*x,y/10)
0.1*Mod(50*x, y)
>>> Mod(5*x,y/10)
Mod(50*x, y)/10
```

This is accomplished with the following (though I don't know if there are other ramifications):

```diff
diff --git a/sympy/core/mod.py b/sympy/core/mod.py
index 23bdd33..cbdae35 100644
--- a/sympy/core/mod.py
+++ b/sympy/core/mod.py
@@ -109,7 +109,7 @@ def doit(p, q):
             cp, p = p.as_coeff_Mul()
             cq, q = q.as_coeff_Mul()
             ok = False
-            if not cp.is_Rational or not cq.is_Rational:
+            if not cp.is_Integer or not cq.is_Integer:
                 r = cp % cq
                 if r == 0:
                     G *= cq
```

certik commented:
Thanks @smichr. That indeed fixes it for the cases I reported above, but with the change in this PR, the floating point case still doesn't work:
```
In [1]: Mod(5*x, y/10)
Out[1]: 
Mod(50⋅x, y)
────────────
     10     

In [2]: Mod(5.*x, y/10)
Out[2]: 0.1⋅Mod(50⋅x, y)

In [3]: Mod(5.*x, y*0.1)
Out[3]: Mod(5.0⋅x, 0.1⋅y)

In [4]: Mod(5.0*x, 0.1*y) == 0.1*Mod(50*x, y)
Out[4]: False
```
I'll have to debug it more.

Fortunately, this seems to be the only failure in this PR.

certik committed:
Temporarily disable a failing test

To fix this test requires modifications in the `Mod` class, which we will do in
subsequent commits and enable the test again.

smichr commented:
Perhaps the following:

```diff
diff --git a/sympy/core/mod.py b/sympy/core/mod.py
index 23bdd33..cefbf8c 100644
--- a/sympy/core/mod.py
+++ b/sympy/core/mod.py
@@ -84,7 +84,7 @@ def doit(p, q):

         # extract gcd; any further simplification should be done by the user
         G = gcd(p, q)
-        if G is not S.One:
+        if G != 1:
             p, q = [
                 gcd_terms(i/G, clear=False, fraction=False) for i in (p, q)]
         pwas, qwas = p, q
```

certik committed:
Fix the Mod(5.0*x, 0.1*y) == 0.1*Mod(50*x, y) test

We fix the Mod class to simplify properly. Thanks @smichr for the patch.

Now the test passes. This change makes another test fail, which we fix in the
next commit.

certik commented:
That works! Another test fails now:
```
In [1]: str(Mod(.6*x + y, .3*y)) == str(Mod(0.1*y + 0.6*x, 0.3*y))
Out[1]: False

In [2]: str(Mod(.6*x + y, .3*y))
Out[2]: 'Mod(0.6*x + 0.1000000000000001*y, 0.3*y)'

In [3]: str(Mod(0.1*y + 0.6*x, 0.3*y))
Out[3]: 'Mod(0.6*x + 0.1*y, 0.3*y)'
```
But I think in this case, the test is incorrect ---- when comparing floating points, we cannot compare them directly, or using their floating point decimal representation, but rather we need to use the `abs(a-b) < eps` idiom. In here, [2] and [3] is equivalent, or to be more precise:
```
In [4]: abs(0.1000000000000001 - 0.1)
Out[4]: 9.71445146547e-17

In [5]: abs(0.1000000000000001 - 0.1) / 0.1
Out[5]: 9.71445146547e-16
```
so we can see that the relative accuracy is 1e-15, you can't get better than that with double precision.

certik committed:
Make the Mod test work

The failing test was the behaving as follows:

In [1]: str(Mod(.6*x + y, .3*y)) == str(Mod(0.1*y + 0.6*x, 0.3*y))
Out[1]: False

In [2]: str(Mod(.6*x + y, .3*y))
Out[2]: 'Mod(0.6*x + 0.1000000000000001*y, 0.3*y)'

In [3]: str(Mod(0.1*y + 0.6*x, 0.3*y))
Out[3]: 'Mod(0.6*x + 0.1*y, 0.3*y)'

As one can see, [2] and [3] are actually equivalent, with more than 1e-14
accuracy in the coefficients. When comparing floating point numbers, one must
use the 'abs(x-y) < eps' idiom. We fix the test by using this idiom for testing that the two expressions are equal when x=y=1.

The only way to make it more robust is to not substitute for the symbols, but
rather extract all coefficients and make sure each one is smaller than 'eps'.
However, we should rather implement some function that does that, so for now I
think the above solution is good enough.

certik committed:
Add a comment explaining what the code does

certik committed:
Add a test for #9398

We test the failure on simple cases first, and then on the actual problem that
was reported.

certik committed:
Add a test for construct_domain() precision

certik commented:
I've added thorough tests and all tests should pass now.


However, while doing that, I discovered that my fix doesn't fix the real issue, consider this:
```
In [2]: simplify(1e-20)
Out[2]: 1.00000000000000e-20

In [3]: simplify(1e-20*I)
Out[3]: 0
```
This manifests already in the `construct_domain()` function, see the test I added in 4c1a2b6ae47a24444063bb6d02399bbe29ed4117, which fails for smaller numbers.

We need to fix it for 1e-20 or any other double precision number.

certik committed:
Make construct_domain() work for higher accuracy

certik committed:
Add higher precision tests for simplify/cancel

certik commented:
Ok, it now works for smaller numbers as well, tests seem to pass.

The actual patch is pretty ugly. It turns out that to represent 1e-20, which fits easily into double precision (53 bits), requires more than 53 to represent --- essentially it requires as many bits as to represent the number 1 + 1e-20. I don't understand why that would be, as that's not how floating point works --- it's called floating point precisely so that you can "float" the point.

But at least this PR now fixes the issue, passes all tests and the tests seem to be thorough. But it would be nice not to require such high accuracy for things like 1e-20.

certik commented:
But it fails if we pass in Float directly to represent even smaller numbers:
```
In [11]: Float("1e-300")
Out[11]: 1.00000000000000e-300

In [12]: Float("1e-500")
Out[12]: 1.00000000000000e-500

In [13]: construct_domain([Float("1e-300")])
Out[13]: 
(RR,
 [1.0000000000000000250590918352087596856961468077037052499253423199004660431840514846763028121819501008949623062702782541489103114649988041308122460916061901827194266279345842755104147827870150702226392606037936139243597750940301438661414791255135908825910173416922229212204049186218220291556195418594185e-300])

In [14]: construct_domain([Float("1e-500")])
Out[14]: (RR, [0.0])
```
Here everything is correct except [14]. The reason is that I only use the `log` trick for numbers down to 1e-323, which was an arbitrary small number I chose for double precision. Still I think it's better than nothing.


To have a general fix, we should remove the hack, and simply use the `_prec` flag, as I was using before, which seems to be correct, and then fix RealField to actually work properly. It looks like the bug is there, that it somehow converts to the number of digits behind the decimal point.

Another bug that is still there is:
```
In [1]: RR(1e-13)
Out[1]: 1.0e-13

In [2]: RR(1e-14)
Out[2]: 0.0

In [3]: RR(1e-20)
Out[3]: 0.0
```
Which again shows, that the actual place to fix this is inside this `RR=RealField`.

certik committed:
Move the fix from construct_domain() into RR

certik committed:
Add tests for RR and double

certik committed:
Use exactly the maximum precision

Fix a test, that now returns a different, but valid result.

certik committed:
Add tests for RR and Float

certik committed:
Add a test for construct_domain() and Float

certik committed:
Update the comment

certik commented:
Ok, now I test all the code changes (and all of the code changes are necessary, otherwise some of the new tests fail) and the changes now actually make sense to me, so I am ok with them.

There are some minor failures in other modules, which I'll try to fix soon.

certik commented:
There is another unrelated bug in Float construction:
```
In [1]: from sympy.polys.constructor import construct_domain

In [2]: f2 = Float("1.0000000000000000000001")

In [3]: construct_domain([f2])
Out[3]: (RR, [1.0000000000000000000001])

In [5]: construct_domain([f2])[1][0]
Out[5]: 1.0000000000000000000001

In [6]: Float(construct_domain([f2])[1][0])
Out[6]: 1.00000000000000

In [7]: Float(construct_domain([f2])[1][0])._prec
Out[7]: 53

In [8]: f2._prec
Out[8]: 80
```
Somehow the last Float lost precision.

smichr commented:
> Somehow the last Float lost precision.

`Float.__new__` doesn't know to watch for this type. Perhaps the simplest thing to do is convert it to a string. A little reordering will make sure that *anything* that isn't recognized is changed to a string:

```diff
diff --git a/sympy/core/numbers.py b/sympy/core/numbers.py
index d1e4f02..f566f82 100644
--- a/sympy/core/numbers.py
+++ b/sympy/core/numbers.py
@@ -664,18 +664,20 @@ class Float(Number):
     is_Float = True

     def __new__(cls, num, prec=None):
-        if isinstance(num, string_types):
-            num = num.replace(' ', '')
-            if num.startswith('.') and len(num) > 1:
-                num = '0' + num
-            elif num.startswith('-.') and len(num) > 2:
-                num = '-0.' + num[2:]
-        elif isinstance(num, float) and num == 0:
+        if isinstance(num, float) and num == 0:
             num = '0'
         elif isinstance(num, (SYMPY_INTS, Integer)):
             num = str(num)  # faster than mlib.from_int
         elif isinstance(num, mpmath.mpf):
             num = num._mpf_
+        else:
+            if not isinstance(num, string_types):
+                num = str(num)
+            num = num.replace(' ', '')
+            if num.startswith('.') and len(num) > 1:
+                num = '0' + num
+            elif num.startswith('-.') and len(num) > 2:
+                num = '-0.' + num[2:]

         if prec is None:
             dps = 15
```

certik committed:
Disable hanging test

certik commented:
Cool, thanks. 

I noticed the tests hang at:
```
sympy/solvers/tests/test_solvers.py[93] ......ff
```
It was an XFAILed test anyway, so I just skipped it in the last commit.

certik committed:
Add Float test to simplify() and cancel()

certik committed:
Fix a test in test_medium

In master it behaves as follows:

    In [1]: from sympy.physics.optics import Medium
    In [2]: from sympy.physics.units import c, u0, e0, m, kg, s, A
    In [3]: m5 = Medium('m5', permittivity=710*10**(-12)*s**4*A**2/(m**3*kg), n=1.33)

    In [4]: m5.intrinsic_impedance - 6.24845417765552*kg*m**2/(A**2*s**3)Out[4]:
                            2
    3.5527136788005e-15⋅kg⋅m
    ─────────────────────────
               2  3
              A ⋅s

    In [5]: simplify(_)
    Out[5]: 0

The result [5] is incorrect. In this PR, it behaves identically, except:

    In [5]: simplify(_)
    Out[5]:
                            2
    3.5527136788005e-15⋅kg⋅m
    ─────────────────────────
               2  3
              A ⋅s

Which is correct. We fixed the test to test that the numerical coefficient is
small enough.

We also fixed the other failing tests.

certik committed:
Fix the test_issue_6828() test

Previously it was using .equal() to compare two expressions containing floating
point, which is incorrect --- .equal() is using symbolic manipulations to see
if the two expressions are equal, which cannot work for a floating point. The
only way to compare expressions that contain floating point is to substitute
random numbers for the symbols and making sure the (numerical) answer is close
enough to zero.

We use verify_numerically() to do this.

certik commented:
@smichr I created a new PR #9415 with your patch, as it introduces some failures and it is not really relevant to the current PR, which I would like to merge soon. We can work on this problem in #9415.

As far as the current PR goes, I think I fixed all failures and added a few more tests. Now I am waiting for Travis to finish testing. If all tests pass, the it is ready for a final review and a merge.

certik committed:
Remove unnecessary import

certik commented:
Travis seems to have some problems, the Theano tests failed to install scipy (I restarted the tests). Otherwise everything else passes.

@smichr, @TheLartians would you both mind reviewing this pull request?

@TheLartians, it should fix the problem that you reported. After it's reviewed, we can merge it.

TheLartians commented:
Well, I don't know anything about sympy internals, however I tested it and the fix seems to work fine on my Systems. Thanks!

smichr commented on sympy/core/tests/test_arit.py:
```diff
@@ -1519,7 +1519,14 @@ def test_Mod():
     assert (x + 3) % 1 == Mod(x, 1)
     assert (x + 3.0) % 1 == Mod(1.*x, 1)
     assert (x - S(33)/10) % 1 == Mod(x + S(7)/10, 1)
-    assert str(Mod(.6*x + y, .3*y)) == str(Mod(0.1*y + 0.6*x, 0.3*y))
```
Any reason to not use the str test?

smichr replied:
Or `assert all(verify_numerically(i, j) for i, j in zip(a.args, b.args))`

certik replied:
The reason why `str` cannot be used to compare (any) floating point numbers is that you need to use the `abs(x-y) < eps` idiom, since two floating point numbers `x` and `y` can be *equal* for the given accuracy `eps`, but their string representation can be different, e.g. 3.999 and 4 are equal for accuracy 1e-3. If you do `x==y` you are using machine precision for `eps`, which is too tight, since any arithmetic operation loses accuracy. The `str` comparison is using whatever accuracy `str` prints, which you have no easy control over (this is set at other places in the code), and it can still print 3.999 instead of 4 for 1e-3 accuracy, so it's not a good way either. 

Conclusion: the only way things can be compared is using the `abs(x-y) < eps`, which is exactly what I do below for this test.

To answer your second question, yes, that would probably work too.

smichr commented on sympy/physics/optics/tests/test_medium.py:
```diff
@@ -29,8 +29,11 @@ def test_medium():
-    assert simplify(m5.permittivity - 7.1e-10*A**2*s**4/(kg*m**3)) == 0
-    assert simplify(m5.permeability - 2.77206575232851e-8*kg*m/(A**2*s**2)) == 0
+    assert abs(m5.refractive_index - 1.33000000000000) < 1e-12
+    assert abs(m5.permittivity - 7.1e-10*A**2*s**4/(kg*m**3)) \
```
Could you use verify numerically again: `assert verify_numerically(m5.permittiity.as_coeff_Mul()[0], 7.1e-10)` and similar for the following? Mostly nits...if you don't want to that's fine.

certik replied:
I can try, if it works. I would suggest to do this in another PR, once this is merged, since I think the current solution is good enough and tests pass.

Btw, we need to greatly improve `verify_numerically` using code from #8036.

smichr commented on sympy/polys/domains/mpelements.py:
```diff
@@ -22,10 +22,7 @@ def _set_mpf(self, val):
-            self.__mpf__ = fzero
-        else:
-            self.__mpf__ = val
+        self.__mpf__ = val
```
I'm not sure that one *doesn't* want numbers to go to zero based on the current precision and rounding settings. It seems like this sort of thing would be desired when solving matrix systems containing number of this type.

certik replied:
So as seen in this PR, sometimes you want it to go to zero and sometimes you don't. In particular, a floating point number, let's say double precision, can hold 15 digits, but the mantissa is arbitrary (for machine doubles, the mantissa integer must fit within 11 bits, for mpf I think there is no limitation). As such, the precision is relative precision. You can easily represent numbers like 1.2345e-100. No problem:
```
In [1]: 1.2345678901234567e-100
Out[1]: 1.2345678901234567e-100

In [2]: _ * 2
Out[2]: 2.4691357802469134e-100

In [3]: _ / 2
Out[3]: 1.2345678901234567e-100

In [4]: _ * 3
Out[4]: 3.70370367037037e-100

In [5]: _ / 3
Out[5]: 1.2345678901234567e-100
```
It holds ~15 digits. You do not want this to become 0, only because the number is smaller than 1e-16:
```
In [6]: 1.2345678901234567e-100 < 1e-16
Out[6]: True
```
and neither we want in sympy, because then it loses the digits as reported in #9398, which this PR fixes.

In applications where you really want this behavior, then you just need to "chop" these small numbers in higher level code. As you can see, all tests pass in sympy with this PR, so we do not actually need this feature (yet). But if we do, you just chop it.

On the other hand, if you always chop it, then you are losing precision and you can't recover it later once you chop it. 

For that reason, it should not be chopped at this low level, but only at higher level, if needed.

certik replied:
Note: this particular truncation code was introduced in https://github.com/sympy/sympy/commit/d191d9379f45ca472b018c928a0e3954ee6efee3#diff-5cab79e07eace0f7c37ec2a8d2f131c6R19 as part of #2126.

smichr replied:
The docstring at the top of the file says "Real and complex elements with built-in truncation." My only fear is that although the change you have made fixes an issue, there might be a lot more to give attention to: 1) the unused `prec`, `rounding` and `tol` are orphaned variables now (that should probably be deleted); 2) the docstring at the top of the file should reflect what is actually happening in the file; and 3) perhaps the same changes should be applied to the lines at 39 and following (which deal with complex numbers). 

I'm just not sure that the change is commensurate with the amount of thought that needs to go into making sure that this is what is intended by this file's purpose.

> It holds ~15 digits. You do not want this to become 0, only because the number is smaller than 1e-16

Yes, I think you are right here. Should *any* sort of truncation (perhaps in digits) be happening in this file?

certik replied:
@smichr those are good points that will require further refactoring. I would delete any kind of truncation.

> Should any sort of truncation (perhaps in digits) be happening in this file?

I don't think so. But I didn't write this file, so I am not sure what the purpose was. @mattpap, what was the motivation behind truncating the numbers? Is there any reason to do so?

mattpap replied:
@certik, yes, truncation to zero is needed to get degree reduction in polynomial division algorithms. The current approach is most likely wrong (as I recall), but nevertheless, there must be some way to allow polynomial division in `RR`.

certik replied:
@mattpap how can I test that the polynomial division is not broken? Because all tests pass with my change.

certik replied:
@smichr I have addressed your objections 1), 2) and 3).

smichr replied:
A few tests for complex numbers like those for the real will guard the new changes that you have made from future regression. Otherwise, I guess with any evidence to the contrary except mattpap's comments, I am +1.

certik replied:
I've added the test in 0018aeb63ce5697f46c29b5018ebfab11fd8324e.

certik commented:
@smichr thanks for the additional comments. The minor comments in tests should be addressed in further PR. As to your important conceptual comment https://github.com/sympy/sympy/pull/9403#discussion_r30981887, I think I addressed it above. But let me know if you have any doubts.

Let's get this merged soon.

certik commented:
@smichr I don't want to delay this any further, as it fixes an important bug. Can you give it +1 or reply to my comments above?

TheLartians commented:
On a side note, may I ask what advantage is gained by using floats internally? To me it seems a bit redundant (and in some extreme cases even a waste of memory), since we can already reach any rational number easily using only fractions and power functions of integers and thereby not having to deal with any precision errors. Only case I can think of is situations where fast numerical evaluation is important, but for this there already exsists lambdify and similar conversions.

certik commented:
@TheLartians in the bug #9398 that you reported, you pass in a float. So SymPy then has no other choice than to use it and do its best.

If you want to avoid such issues, then use rational numbers, as you said, and then there should be no problem.

TheLartians commented:
I would have expected that floats will be directly converted to rational numbers (using the maximum python float accuracy). That way absolute precision will be guaranteed or am I mistaken?

certik commented:
@TheLartians yes, it would. You can do that yourself before passing it in.

TheLartians commented:
What I was wondering is, would this provide any disadvantages over the current approach?

toolforger commented:
Am 27.05.2015 um 23:23 schrieb Lars Melchior:
> I would have expected that floats will be directly converted to rational numbers (using the maximum python float accuracy).

We do support floats directly, with the usual rounding semantics.
Reason is that if you do rationals, the numerator and denominator will 
become longer and longer.
Runge-Kutta and similarly iterative approaches will quickly become 
unmanageable with that. Symbolic math usually doesn't have iterative 
algorithms, so it can work.

 > That way absolute precision will be guaranteed or am I mistaken?

It will, but be aware that the floats that you input typically have an 
error margin.
Also, 0.5 will probably be what you want, but 1.0/3.0 will not because 
1/3 cannot be expressed precisely as a float. If you have that kind of 
number in your inputs, you are far better off constructing 
`Rational(1,3)`, or doing `S(1)/3` (`S(1)` returns an exact type from 
SymPy, and arithmetic involving an exact SymPy type will always try to 
give you an exact result.

TheLartians commented:
Ok I see. However I think it is also problematic to use only rationals for exact values. For example, entering a very large number such as `x = S(10)**S(100000000)` will cause sympy to freeze (I'm guessing since it will attempt to convert the number to a rational by appending lots of zeros). I would rather keep the exponent, so basically implement floating point in sympy. This would keep all the advantages of the current adaptive precision approach for floats without having to deal with precision issues and having two separate systems for numbers.

certik committed:
Delete unused variables

certik committed:
Remove truncation from _set_mpc()

certik committed:
Elements are not truncated anymore

smichr commented:
> will work and instantly ...  take a very long time to evaluate

You are seeing the effects of caching. The first expression involving 10**i*10**-i (for i a large number) *whether there is another symbol as a factor or not* will take a long time but will be "instantaneous" the next time since the result is known.

TheLartians commented:
@smichr Unfortunately this is not true. the first expression will evaluate very quickly regardless of previous input (and in more or less constant time regarding the exponent), while the second will take a incredibly long time to evaluate (and caching will not work if he exponents are changed). FYI I created a quick plot of the average time needed to evaluate these expressions for different exponent lengths [here](http://nbviewer.ipython.org/gist/TheLartians/e36ffbbe86113e779d65).
The difference between the expressions is, that in the first expression the exponent is multiplied with a "dummy symbol" `Symbol("1")` which makes it a symbolic expression that cannot be reduced to a rational. Hence the huge speed difference.
Note that passing python floating points here will also work without big delay, however imho it is impractical for end-users to need to think about two separate systems for numbers depending on case.
Sorry for not making myself clearer. My point is that

- replacing rational numbers with expressions like `a/b * 10**e` with shortest possible integers `a`,`b` and integer `e` and
- handling python floats by converting them to such expressions (e.g. `0.123 = 123 * 10**-3`)

would be able to solve booth the big number problems and the precision problem elegantly.

smichr commented:
> Unfortunately this is not the case here

I agree. I used expressions where the symbol was a factor, not a factor in the exponent.

Regarding the creation of Floats from strings. Yes, the precision is locked in at the creation (and you can create .3 with 50 digits of precision if you want using Float('.3', 50)) but you are right that we don't have an option for representing infinite zeros past the digits given except to do so as a Rational (e.g. Rational('0.3') is exactly 3/10). I can see the wisdom in converting rationals to expressions as you have described if one wants to efficiently handle incomprehensibly large numbers, but can't foresee the ramifications of doing so for Floats which are inherently of limited precision.

smichr commented:
> replacing rational numbers with expressions like a/b * 10**e

The underlying mpf values with which SymPy works are stored as `a*2**b` so if we did as you are saying it would mean a rewrite of how numbers are handled.

TheLartians commented:
> The underlying mpf values with which SymPy works are stored as a*2**b so if we did as you are saying it would mean a rewrite of how numbers are handled.

As long as we use only integers (so b is always positive) there is nothing wrong storing them that way, since we can represent every integer exactly. However with floating-point numbers this is not possible anymore (as in the example I provided, 0.2 cannot be represented accurately with finite binary digits). Thats why floating point needs to be implemented in sympy itself and cannot be stored in libraries using binary numbers internally. 

I don't think this is much work, note that using the trick with the dummy symbol already works in many cases. Only the implicit conversion to rationals has to be replaced by rational times exponent and the conversion of floating points to this type of representation. What doesn't work yet, however, are simple divisions in the form`a**b/a` such as
``S(10)**(S(-10)*Symbol('1'))/S(10)``,
which sympy should reduce to
``10**(-10*1 - 1)``,
but does nothing at the moment. I believe this is an easy to implement and much required feature anyways.

TheLartians commented:
> but can't foresee the ramifications of doing so for Floats which are inherently of limited precision.

For example, the current method is always prone to precision errors (just imagine a product of any number with 0.2). My method would fix this.

smichr commented:
> somehow only this exact case

As I recall, not all possible base combinations are tested for speed reasons. One can use simplification rules like powsimp, however, to make those transformations.

Regarding the 'How about' expression, we don't convert string expressions to the desired precision, just string numbers as in

```python
>>> (pi/10).n(25)
0.3141592653589793238462643
>>> (pi*Float('.1',25)).n(25)
0.3141592653589793238462643
```

I'm not trying to be antagonisitic. If you think you have a better way to represent numbers, ask on the mailing list for feedback and if it's positive you might want to code up a PR.

TheLartians commented:
> As I recall, not all possible base combinations are tested for speed reasons. One can use simplification rules like powsimp, however, to make those transformations.

Well it seems here that not only the base but also the exponents are tested since only this particular case works, I imagine that testing only for identical base and applying `a**b*a**c = a**(b+c)` would be faster. But thats another issue.

> If you think you have a better way to represent numbers, ask on the mailing list for feedback and if it's positive you might want to code up a PR.

I think I will do this. Thanks for your input!

certik committed:
Merge branch 'master' into issue9398

smichr commented:
@certik, I gave feedback you might not have noticed [here](https://github.com/sympy/sympy/pull/9403#discussion_r31367805)

certik committed:
Add a test for CC(1e-40) and CC(1e-40j)

This tests the change introduced in 947538f43c80589b1f79448f4da2409b4c53eb87.

certik commented:
@smichr thanks, I did notice it, but only now I got to actually implement it.

Is the PR +1 from you to merge as it is, if tests pass?

smichr commented:
I'm +1 if the polynomial division (that @mattpap referred to) still works. But since there are no tests that fail and your use case (identified by an issue) is driving this change, I guess it's good to fix the existing problem. If I were looking at this I know I would need a big whiteboard to be able to think out what should be happening with precision and accuracy in this context.

certik commented:
Tests pass, @smichr gave +1, so merging.
